app/src/main/cpp/native-lib.cpp

```diff
@@ -12,9 +38,203 @@ Java_com_example_cloudonix_MainActivity_stringFromJNI(
 }
 
 JNICALL
-Java_com_example_cloudonix_MainActivity_stringFromJNI(
+extern "C"
+jstring Java_com_example_cloudonix_MainActivity_getifaddrs(
         JNIEnv *env,
         jobject /* this */) {
-    std::string hello = "Hello from C++";
-    return env->NewStringUTF(hello.c_str());
+
+    jstring retVal = nullptr;
+    // To make things simple, this method **will not** instantiate Java objects
+    // and return an int value. Instead, it will only return a jstring.
+    //
+    // It will first invoke the necessary getifaddrs() native method,
+    // sift through it's returned values, filtering out what's interesting to
+    // return, and just return that one address as a string. It won't bother
+    // converting the ifaddr structs to actual classes or whatnot.
+    //
+    // Moreover, it will very crudely check whether the address is actually a private
+    // address. It will also presume that, if the IP starts with "10." that it belongs to
+    // a 10.x.x.x subnet without actually checking whether the IP is valid or not.
+
+    /**
+     * 1. If an IPv6 address is available, that is from the global unicast range - as per this document, return that address.
+     * 2. If an IPv4 address is available that is public - i.e. not a “private address” according to the table here, or “link local” according to the specification here, return that address.
+     * 3. Otherwise, return any one of the IPv4 addresses available.
+     */
+
+    // IPv6 global unicast range addresses start with 2000: so check for them
+
+    /**
+     * IPv4 private addresses can be one of the following three kinds:
+     * 1. 10.0.0.0 - 10.255.255.255
+     * 2. 172.16.0.0 - 172.31.255.255
+     * 3. 192.168.0.0 - 192.168.255.255
+     */
+
+    /**
+     * IPv4 link local addresses look like this:
+     * 169.254.0.0 – 169.254.255.255
+     */
+
+    // First things first, fire the JNI call
+    struct ifaddrs *ifaddr, *ifaddr_iterator;
+    char addr[INET6_ADDRSTRLEN];
+
+    if (getifaddrs(&ifaddr) == -1) {
+        perror("getifaddrs error");
+    }
+
+    for (ifaddr_iterator = ifaddr;
+         ifaddr_iterator != NULL; ifaddr_iterator = ifaddr_iterator->ifa_next) {
+        // Iterate IPv6 addresses first
+        if ((ifaddr_iterator->ifa_addr) && (ifaddr_iterator->ifa_addr->sa_family == AF_INET6)) {
+            struct sockaddr_in6 *in6 = (struct sockaddr_in6 *) ifaddr_iterator->ifa_addr;
```

Comment:
This is still unused. I dont know where I was going with this, but i'll leave it hanging for the time being... maybe i end up needing it during debugging.

leave a `//TODO remove me` before merging in please.

Reply:
done. resolving.